python/pyspark/sql/dataframe.py

```diff
@@ -2153,6 +2153,22 @@ def transform(self, func):
                                               "should have been DataFrame." % type(result)
         return result
 
+    @since(3.0)
+    def sameSemantics(self, other):
+        """
+        Return true when the query plan of the given :class:`DataFrame` will return the same
+        results as this :class:`DataFrame`.
+        """
+        return self._jdf.sameSemantics(other)
```

Comment:
Since Python API side does not check the type, we could add an if-else with throwing `ValueError`.

```python
if not isinstance(other, DataFrame):
    raise ValueError("other parameter should be of DataFrame; however, got %s" % type(other))
```